src/main/java/com/cdancy/jenkins/rest/parsers/OptionalFolderPathParser.java

```diff
@@ -0,0 +1,32 @@
+package com.cdancy.jenkins.rest.parsers;
+
+import com.cdancy.jenkins.rest.JenkinsUtils;
+import com.google.common.base.Function;
+
+import javax.inject.Singleton;
+
+/*
+ * Turn the optionalFolderPath param to jenkins URL style if needed
+ */
+@Singleton
+public class OptionalFolderPathParser implements Function<Object,String> {
+
+    public static final String EMPTY_STRING = "";
+    @Override
+    public String apply(Object optionalFolderPath) {
+        if(optionalFolderPath == null) {
+            return EMPTY_STRING;
+        }
+
+        String folderPath = String.class.cast(optionalFolderPath);
+        String amendedPath = "";
+        if(folderPath.startsWith("/") || folderPath.endsWith("/")) {
+            throw new RuntimeException("Incorrect folder Path format - " + folderPath);
+        }
+        String[] folderNames = folderPath.split("/");
+        for (String folder:folderNames) {
+            amendedPath += "job/" + folder + "/";
```

Comment:
I wonder about this: what happens if the user puts in a non-simple path? Meaning the fully qualified `job/hello/job/world/job/build`.

Reply:
Like @martinda pointed, It is hard to say if "job/job" is a fully qualified path to a job or simple folder path(Apparently, folders can also be named as "job"). I am not sure if we can check for a fully qualified path before we amend it.

When users say `job/hello/job/world/job/build`  this logic will go ahead add `job/` before every word and probably would return a 404 not found status.

Reply:
In the fully qualified job name `job/hello/job/world/job/build`, there are two things to consider:
1. The path (hello/world) - forward slashes used for nesting if necessary.
1. The job name (build) - forward slashes not allowed.

This pattern reminds me of the Java [`File(String parent, String child)`](https://docs.oracle.com/javase/8/docs/api/java/io/File.html#File-java.lang.String-java.lang.String-) constructor, for which there are effectively multiple constructors. It also reminds me of the [`createDirectories`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#createDirectories-java.nio.file.Path-java.nio.file.attribute.FileAttribute...-) in nio Files.

Let's imagine for a moment that there is an API to create a folder path, and that it returns a `JenkinsFolder` instance:

```java
    JenkinsFolder createFolders(String folderName)
```
That api would create folders on Jenkins, thereby validating the user input (simple path expression) and returning some read-only instance representing that path.

Then there would be two APIs to create a job:
1. `create(JenkinsFolder folder, String jobName)`
2. `create(String jobName)`

Regarding the path, the regex to detect a fully qualified path is: `Pattern pattern = Pattern.compile('(job/[^/]+?)(/job/[^/])*');`. However it fails if the user is "crazy" enough to call a folder "job" and a nested folder "job", because in this case it is ambiguous.

Ultimately, only Jenkins can completely validate the `folderPath` and the `jobName`. IMO it is fine to let Jenkins return 404 if the user incorrectly specifies them. I would prefer that jenkins-rest does not interfere with what Jenkins can accept.

I also prefer a solution that is predictable from the user point of view. If the internals contain different behaviors that depend on the input value, those need to be explained in the wiki. Potentially, the more internal manipulations the jenkins-rest does, the more difficult it is for the user to debug.

I think a lot... perhaps too much... what do you think?

Reply:
I have not considered all the cases. Specifically, the case where a `JenkinsFolder` already exists. In this case, `createFolder(String folerName)` is not the right call to create an instance of `JenkinsFolder`.

Reply:
@martinda so is the thought here that users can ONLY pass in the "simple names"? Would it make sense to check each iteration and if it's not preceded with "job" then insert it otherwise assume OK? I'd like to support both the simple case and the normal case if possible.

Reply:
Let's assume:
1. simple names, simple case: `a/b/c`
1. normal names, normal case: `job/a/job/b/job/c` (aka fully qualified folder)

> so is the thought here that users can ONLY pass in the "simple names"?

Yes.

> Would it make sense to check each iteration and if it's not preceded with "job" then insert it otherwise assume OK?

Yes but there is an ambiguity: a user can create a folder called "job". Unfortunately, this is possible (I tried it). Because of the ambiguity, the code will not work if a user creates nested folders called "job" (this would be strange, but Jenkins allows it).

>I'd like to support both the simple case and the normal case if possible.

Me too. But I do not know what to do about the ambiguity. So in my mind the code should only support one of the two cases. I am good with either one.

Reply:
@martinda while I though there must be _something_ we can do, after playing with things for a while, I see that its impossible given that a job name can actually be `job`. If it weren't for this we'd be able to check but it's simply impossible to code for things if `job` is a valid name.